plugins/techdocs-backend/src/service/CachedEntityLoader.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { CatalogClient } from '@backstage/catalog-client';
+import { CacheClient } from '@backstage/backend-common';
+import {
+  Entity,
+  EntityName,
+  stringifyEntityRef,
+} from '@backstage/catalog-model';
+import { IdentityClient } from '@backstage/plugin-auth-backend';
+import { ResponseError } from '@backstage/errors';
+
+export type CachedEntityLoaderOptions = {
+  catalog: CatalogClient;
+  cache: CacheClient;
+  identity: IdentityClient;
+};
+
+export class CachedEntityLoader {
+  private catalog: CatalogClient;
+  private cache: CacheClient;
+  private identity: IdentityClient;
+
+  constructor({ catalog, cache, identity }: CachedEntityLoaderOptions) {
+    this.catalog = catalog;
+    this.cache = cache;
+    this.identity = identity;
+  }
+
+  async load(
+    entityName: EntityName,
+    token: string | undefined,
+  ): Promise<Entity | undefined> {
+    const cacheKey = await this.getCacheKey(entityName, token);
+    let result = (await this.cache.get(cacheKey)) as Entity | undefined;
+
+    if (result) {
+      return result;
+    }
+
+    try {
+      result = await this.catalog.getEntityByName(entityName, { token });
+    } catch (err) {
+      if (err instanceof ResponseError && err.response.status === 403) {
+        result = undefined;
+      } else {
+        throw err;
+      }
+    }
+
+    if (result) {
+      await this.cache.set(cacheKey, result, { ttl: 5000 });
+    }
+
+    return result;
+  }
+
+  private async getCacheKey(
+    entityName: EntityName,
+    token: string | undefined,
+  ): Promise<string> {
+    const entityRef = stringifyEntityRef(entityName);
+
+    if (!token) {
+      return entityRef;
+    }
+
+    const response = await this.identity.authenticate(token);
```

Comment:
Q: does this also make network calls to the auth backend, or is this just pulling the identifier from the token in code?

Reply:
It pulls the data out of the token, but in the worst case it might result in a call to the auth backend to refresh the key store. See https://github.com/backstage/backstage/blob/a426ba0e26f3596126913c6793cf53eaaa3b25e3/plugins/auth-backend/src/identity/IdentityClient.ts#L131

Reply:
Got it.  And if we were instead to use some derivation of the raw token value itself, is the worst-case just that sometime in the 5s TTL of the cached value, the token changes on the client-side, a new token gets passed instead, and there's a cache miss?  Or are there other implications?

Reply:
Yup! Unless there's some security concern with using tokens in the cache key..

Reply:
Yeah both seem unlikely and should still respond successfully, just a bit slower. We might as well simplify things and remove the identity client dependency